### PR TITLE
Davydden tensor updates

### PR DESCRIPTION
**Pull request by bangerth**
This is an update to #491. It only adds the missing copy constructors to the specializations of class Tensor and updates the test.

### CODE REVIEW

File: include/deal.II/base/tensor.h [hunk 1]
```diff
@@ -103,6 +103,13 @@ class Tensor
   Tensor (const array_type &initializer);
 
   /**
+   * Copy constructor taking a tesnor of another base data type
+   */
```
Comment (kronbichler) [inline include/deal.II/base/tensor.h]: typo. tesnor -> tensor

File: include/deal.II/base/tensor.h [hunk 2]
```diff
@@ -103,6 +103,13 @@ class Tensor
...
+  template <typename OtherNumber>
+  explicit
+  Tensor (const Tensor<rank_,dim,OtherNumber> &);
+
```
Comment (kronbichler) [inline include/deal.II/base/tensor.h]: We often put a dummy name in the declaration to make things more similar to the implementation (but we are inconsequential about it, see below)

File: include/deal.II/base/tensor.h [hunk 3]
```diff
@@ -137,6 +144,9 @@ class Tensor
    */
   Tensor &operator = (const Tensor<rank_,dim,Number> &);
 
+  template <typename OtherNumber>
+  Tensor &operator = (const Tensor<rank_,dim,OtherNumber> &);
+
```
Comment (kronbichler) [inline include/deal.II/base/tensor.h]: Please add the "Assignment operator" comment also here

File: include/deal.II/base/tensor.h [hunk 4]
```diff
@@ -404,6 +414,26 @@ Tensor<rank_,dim,Number>::operator = (const Tensor<rank_,dim,Number> &t)
...
+template <typename OtherNumber>
+inline
+Tensor<rank_,dim,Number> &
+Tensor<rank_,dim,Number>::operator = (const Tensor<rank_,dim,OtherNumber> &t)
+{
+  for (unsigned int i=0; i<dim; ++i)
+    subtensor[i] = t.subtensor[i];
+  return *this;
+}
+
```
Comment (kronbichler) [inline include/deal.II/base/tensor.h]: Three empty lines?

File: include/deal.II/base/tensor_base.h [hunk 5]
```diff
@@ -116,6 +116,14 @@ class Tensor<0,dim,Number>
   Tensor (const Tensor<0,dim,Number> &);
 
   /**
+   * Copy constructor from tensors with different underlying scalar
+   * type. This obviously requires that the @p OtherNumber type is
+   * convertible to @p Number.
+   */
+  template <typename OtherNumber>
+  Tensor (const Tensor<0,dim,OtherNumber> &);
+
```
Comment (kronbichler) [inline include/deal.II/base/tensor_base.h]: Any reason not to make this constructor explicit as for Tensor<rank,dim>? Do we need to allow the conversion float -> Tensor<0,dim,float> -> Tensor<0,dim,double> this way?

File: include/deal.II/base/tensor_base.h [hunk 6]
```diff
@@ -136,6 +144,14 @@ class Tensor<0,dim,Number>
   Tensor<0,dim,Number> &operator = (const Tensor<0,dim,Number> &);
 
   /**
+   * Assignment operator from tensors with different underlying scalar
+   * type. This obviously requires that the @p OtherNumber type is
+   * convertible to @p Number.
+   */
+  template <typename OtherNumber>
+  Tensor<0,dim,Number> &operator = (const Tensor<0,dim,OtherNumber> &);
+
```
Comment (kronbichler) [inline include/deal.II/base/tensor_base.h]: same as above

File: include/deal.II/base/tensor_base.h [hunk 7]
```diff
@@ -370,6 +394,14 @@ class Tensor<1,dim,Number>
   Tensor<1,dim,Number> &operator = (const Tensor<1,dim,Number> &);
 
   /**
+   * Assignment operator from tensors with different underlying scalar
+   * type. This obviously requires that the @p OtherNumber type is
+   * convertible to @p Number.
+   */
+  template <typename OtherNumber>
+  Tensor<1,dim,Number> &operator = (const Tensor<1,dim,OtherNumber> &);
+
```
Comment (kronbichler) [inline include/deal.II/base/tensor_base.h]: and for these two as well

**Comment by kronbichler** [comment]
Apart from these minor issues, it looks good.

**Comment by bangerth** [comment]
All good points -- this is why we do peer review.

I'll fix this later today (the tree is busy with another patchset running tests right now). I interpret your last sentence as approval to merge if changes are made as outlined -- let me know if you meant differently.

**Comment by kronbichler** [comment]
Indeed, that is what I meant!

File: include/deal.II/base/tensor_base.h (hunk 6 above)
Comment (bangerth) [inline include/deal.II/base/tensor_base.h]: Isn't `explicit` on copy operators only C++11?

Comment (bangerth) [inline include/deal.II/base/tensor_base.h]: Ah, forget that -- the change in C++11 is only for conversion operators.

File: include/deal.II/base/tensor_base.h (hunk 7 above)
Comment (bangerth) [inline include/deal.II/base/tensor_base.h]: So I can make the conversion constructor explicit, but assignment operators I can't. So that's what I did.

**Comment by bangerth** [comment]
OK, I've addressed everything.

**Pull request merged**